src/lv_message_efficient.cc

```diff
@@ -117,82 +110,78 @@ namespace grpc_labview
 
     //---------------------------------------------------------------------
     //---------------------------------------------------------------------
-    const char *LVMessageEfficient::ParseBytes(google::protobuf::uint32 tag, const MessageElementMetadata& fieldInfo, uint32_t index, const char *ptr, ParseContext *ctx)
+    const char* LVMessageEfficient::ParseBytes(google::protobuf::uint32 tag, const MessageElementMetadata& fieldInfo, uint32_t index, const char* ptr, ParseContext* ctx)
     {
-        return ParseString(tag, fieldInfo, index, ptr, ctx);    
+        return ParseString(tag, fieldInfo, index, ptr, ctx);
     }
 
     //---------------------------------------------------------------------
     //---------------------------------------------------------------------
-    const char *LVMessageEfficient::ParseNestedMessage(google::protobuf::uint32 tag, const MessageElementMetadata& fieldInfo, uint32_t index, const char *protobuf_ptr, ParseContext *ctx)
-    {    
+    const char* LVMessageEfficient::ParseNestedMessage(google::protobuf::uint32 tag, const MessageElementMetadata& fieldInfo, uint32_t index, const char* protobuf_ptr, ParseContext* ctx)
+    {
         auto metadata = fieldInfo._owner->FindMetadata(fieldInfo.embeddedMessageName);
-        
+
         LVMessageEfficient nestedMessage(metadata);
 
-        if (fieldInfo.isRepeated) {
+        if (fieldInfo.isRepeated)
+        {
             // if the array is not big enough, resize it to 2x the size
             auto numElements = 128;
             auto elementIndex = 0;
             auto clusterSize = metadata->clusterSize;
-            auto alignment = metadata->alignmentRequirement;
             auto arraySize = numElements * clusterSize;
             char _fillData = '\0';
 
             // Get the _repeatedMessageValues vector from the map
-            auto _repeatedMessageValuesIt = _repeatedMessageValuesMap.find(metadata->messageName);
+            auto _repeatedMessageValuesIt = _repeatedMessageValuesMap.find(fieldInfo.fieldName);
             if (_repeatedMessageValuesIt == _repeatedMessageValuesMap.end())
             {
                 auto m_val = std::make_shared<RepeatedMessageValue>(fieldInfo, google::protobuf::RepeatedField<char>());
-                _repeatedMessageValuesIt = _repeatedMessageValuesMap.emplace(metadata->messageName, m_val).first;
+                _repeatedMessageValuesIt = _repeatedMessageValuesMap.emplace(fieldInfo.fieldName, m_val).first;
+                _repeatedMessageValuesIt->second.get()->_buffer.Resize(arraySize, _fillData);
             }
+            else
+            {
+                // Zero out data from the previous use of the buffer.
+                auto bytesUsed = _repeatedMessageValuesIt->second.get()->_numElements * clusterSize;
```

Comment:
NOTE: Rather than using `_continueFromIndex` to begin writing from where the previous element in the array left off, we now always reset the index back to zero. For that to work, we need to explicitly zero out any data from the previous element. Otherwise, if some fields weren't serialized because they were set to their default values, we can end up reusing the data from the previous element which leads to data corruption. I thought about just allocating a new buffer each time to simplify things, but I decided the performance gains of not repeatedly allocation a buffer were probably worth it.

The problem with the previous approach is that while we were effectively tracking the end index of the buffer, we weren't also tracking the beginning index. So when we finally copy the data in `PostInteralParseAction`, we always copy from index zero which includes data from all of the elements that preceded it. Also, the previous approach allocated ever growing memory rather than reuse the memory from previous allocations once it had been copied.